### PR TITLE
Changes to text parser to handle decode errors #3301

### PR DESCRIPTION
**Pull request by joachimmetz**
Changes to text parser to handle decode errors #3301

### CODE REVIEW

File: plaso/parsers/text_parser.py [hunk 1]
```diff
@@ -179,6 +180,33 @@ def __init__(self):
...
+      raise TypeError('Unsupported exception type.')
+
+    if self._parser_mediator:
+      self._parser_mediator.ProduceExtractionWarning(
+          'error decoding 0x{0:02x} at offset: {1:d}'.format(
+              exception.object[exception.start],
+              self._current_offset + exception.start))
+
+    escaped = '\\x{0:2x}'.format(exception.object[exception.start])
+    return (escaped, exception.start + 1)
```
Comment (Onager) [inline plaso/parsers/text_parser.py:209]: Shouldn't this be exception.**end** (no +1) instead? I assume start and end will be the same after a 1-byte error, but if there's multiple bytes of invalid data, wouldn't we want to resume parsing at the end of that?

File: plaso/parsers/text_parser.py [hunk 2]
```diff
@@ -179,6 +180,33 @@ def __init__(self):
...
+    codecs.register_error('text_parser_handler', self._EncodingErrorHandler)
+
+  def _EncodingErrorHandler(self, exception):
+    """Encoding error handler.
+
+    Args:
+      exception [UnicodeDecodeError]: exception.
+
+    Returns:
+      tuple[str, int]: replacement string and number of bytes to skip.
```
Comment (Onager) [inline plaso/parsers/text_parser.py:194]: The docs for [register_error() ](https://docs.python.org/3/library/codecs.html#codecs.register_error) say that the int here is where encoding should continue, not the number of bytes that should be skipped. Please correct the docstring here.

File: plaso/parsers/text_parser.py [hunk 3]
```diff
@@ -179,6 +180,33 @@ def __init__(self):
...
+    if not isinstance(exception, UnicodeDecodeError):
+      raise TypeError('Unsupported exception type.')
+
+    if self._parser_mediator:
+      self._parser_mediator.ProduceExtractionWarning(
+          'error decoding 0x{0:02x} at offset: {1:d}'.format(
+              exception.object[exception.start],
+              self._current_offset + exception.start))
+
+    escaped = '\\x{0:2x}'.format(exception.object[exception.start])
```
Comment (Onager) [inline plaso/parsers/text_parser.py:208]: This only handles the case of a 1-byte error, are you sure you want 1 exception message per invalid byte?

**Review by Onager — CHANGES_REQUESTED** [review]

File: plaso/parsers/text_parser.py (hunk 1 above)
Comment (joachimmetz) [reply plaso/parsers/text_parser.py:209]: I'm not sure, I took the conservative approach to account for single byte data that might case multi byte decoding errors

File: plaso/parsers/text_parser.py (hunk 2 above)
Comment (joachimmetz) [reply plaso/parsers/text_parser.py:194]: changed

File: plaso/parsers/text_parser.py (hunk 3 above)
Comment (joachimmetz) [reply plaso/parsers/text_parser.py:208]: I took the conservative approach to account for single byte data that might case multi byte decoding errors. We could limit the number of warning that will be generated if you expect a lot of encoding errors.

**Comment by joachimmetz** [comment]
@Onager PTAL

**Pull request merged**